src/mlpack/methods/ann/layer/read_memory.hpp

```diff
@@ -0,0 +1,276 @@
+/**
+ * @file read_memory.hpp
+ * @author Sumedh Ghaisas
+ *
+ * Definition of Read Memory layer used in Neural Turing Machine.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_READ_MEMORY_HPP
+#define MLPACK_METHODS_ANN_LAYER_READ_MEMORY_HPP
+
+#include <mlpack/prereqs.hpp>
+
+#include "../visitor/delta_visitor.hpp"
+#include "../visitor/output_parameter_visitor.hpp"
+
+#include "layer_types.hpp"
+
+#include "memory_head.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * An implementation of a Read Memory used in NTM.
+ * The layer takes the controller output as input and computes the read memory
+ * content.
+ *
+ * @tparam InputDataType Type of the input data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ * @tparam OutputDataType Type of the output data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ */
+template <
+    typename InputDataType = arma::mat,
+    typename OutputDataType = arma::mat
+>
+class ReadMemory
+{
+ public:
+  /**
+   * Create the Read Memory layer object using the specified parameters.
+   *
+   * @param inSize The number of input units.
+   * @param numMem Number of memory locations.
+   * @param memSize Memory size in each memory block.
+   * @param shiftSize Circular convolutional shift size used.
+   */
+  ReadMemory(const size_t inSize,
+             const size_t numMem,
+             const size_t memSize,
+             const size_t shiftSize);
+
+  /**
+   * Feed forward pass of a neural network, evaluating the function
+   * f(x) by propagating the activity forward given the current memory content.
+   *
+   * @param input Input data used for evaluating the specified function.
+   * @param output Resulting output activation.
+   * @param memory Current memory content.
+   */
+  template<typename eT>
+  void ForwardWithMemory(arma::Mat<eT>&& input,
+                         const arma::Mat<eT>&& memory,
+                         arma::Mat<eT>&& output);
+
+  /**
+   * Feed forward pass of a neural network, evaluating the function
+   * f(x) by propagating the activity forward given the current memory content.
+   *
+   * This function is used in testing the layer with MemoryTest layer.
+   *
+   * @param input Input data used for evaluating the specified function.
+   * @param output Resulting output activation.
+   * @param memory Current memory content.
+   */
+  template<typename eT>
+  void ForwardWithMemoryTest(arma::Mat<eT>&& input,
+                             const arma::Mat<eT>&& memory,
+                             arma::Mat<eT>&& output)
```

Comment:
ForwardWithMemoryTest is used by MemoryTest layer to call layer while testing the gradients. Although I think I can make some design changes  to counter this problem. Although that would involve deleting ReadMemory and WriteMemory layers and incorporating them into NTM itself. I thought of implementing this design but then decided to go with a more modularized approach. 

Okay so the main reason this function exists is because of WriteMemory layer. This layer accepts a memory, and the controller output and outputs a new memory content. Although for gradient testing this layer is used alone with MemoryTest layer (classic Unit testing... haha :) ). So this function reshapes the output to a vec so that this layer is just like a normal layer rather than being a layer which outputs a memory. This function is thus only called by MemoryTest layer while testing.

To bypass this, we need to make the memory changes in NTM so that WriteMemory just outputs the memory location to make the change. But then its just a MemoryHead ;P So no need of that layer. Let me know what you think.

Ahh and MemoryTest for the layers are done in ann_layer_test.cpp